internal/controlplane/common.go

```diff
@@ -44,6 +45,22 @@ func providerError(err error) error {
 	return fmt.Errorf("provider error: %w", err)
 }
 
+// providerBuilderFactory is a helper function to build a provider builder
+// It has the added value that it will trigger some devs XD
+func (s *Server) providerBuilderFactory(ctx context.Context, provider db.Provider) (*providers.ProviderBuilder, error) {
+	pbOpts := []providers.ProviderBuilderOption{
+		providers.WithProviderMetrics(s.provMt),
+		providers.WithRestClientCache(s.restClientCache),
+	}
+
+	p, err := providers.GetProviderBuilder(ctx, provider, s.store, s.cryptoEngine, &s.cfg.Provider, pbOpts...)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "cannot get provider builder: %v", err)
+	}
+
+	return p, nil
+}
```

Comment:
I'd really love for this to be a struct which matches an interface, e.g.

```golang
struct ProviderClass interface {
	Get(ctx, provider) (ProviderBuilder, error)
}

type githubProvider struct {
	store   db.Querier
	cryptoEngine crypto.Engine
	cfg    *serverconfig.ProviderConfig
	opts   []ProviderBuildOption
}

func (ghp *githubProvider) Get(ctx, provider) (ProviderBuilder, error) {
	return providers.GetProviderBuilder(ctx, provider, ghp.store, ghp.cryptoEngine, ghp.cfg, ghp.opts...)
}
```

We could then store a ProviderClass on Server, which would make it easier to mock the ProviderClass in tests later, or compose a ProviderClass from multiple implementations.

Does that make sense?

Reply:
(I'm also thinking we should define a `ProviderInterface` that `ProviderBuilder` implements, but that's a separate topic.

Reply:
Strongly agree with @evankanderson on this one.

Reply:
Strongly agree with @evankanderson on this one.